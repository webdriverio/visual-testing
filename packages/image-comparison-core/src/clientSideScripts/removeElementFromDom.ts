/**
 * Remove a DOM element from the DOM
 */
export default function removeElementFromDom(id: string): void {
    const elem = document.querySelector(`#${id}`)
    if (elem) {
        elem.parentNode!.removeChild(elem)
    }
}
