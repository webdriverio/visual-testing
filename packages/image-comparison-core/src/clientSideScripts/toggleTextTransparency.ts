/**
 * Add or remove the transparency from the text
 */
export default function toggleTextTransparency(enableTransparency: boolean): void {
    if (enableTransparency) {
        const elements = document.querySelectorAll<HTMLElement>('body *:not(script):not(style)')
        elements.forEach(el => {
            el.style.setProperty('color', 'transparent', 'important')
        })
        return
    }

    const elements = document.querySelectorAll<HTMLElement>('body [color]:not(script):not(style)')
    elements.forEach(el => {
        el.style.removeProperty('color')
    })
}
