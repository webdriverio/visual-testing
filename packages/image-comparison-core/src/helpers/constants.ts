import type { IosOffsets } from './constants.interfaces.js'
import type { ResizeDimensions } from '../methods/images.interfaces.js'
import type { TestContext } from 'src/methods/compareReport.interfaces.js'
import type { DeviceRectangles } from '../methods/rectangles.interfaces.js'

export const DEFAULT_COMPARE_OPTIONS = {
    blockOutSideBar: true,
    blockOutStatusBar: true,
    blockOutToolBar: true,
    createJsonReportFiles: false,
    diffPixelBoundingBoxProximity: 5,
    ignoreAlpha: false,
    ignoreAntialiasing: false,
    ignoreColors: false,
    ignoreLess: false,
    ignoreNothing: false,
    rawMisMatchPercentage: false,
    returnAllCompareData: false,
    saveAboveTolerance: 0,
    scaleImagesToSameSize: false,
}
export const DEFAULT_FORMAT_STRING = '{tag}-{browserName}-{width}x{height}-dpr-{dpr}'
export const STORYBOOK_FORMAT_STRING = '{tag}-{logName}-{width}x{height}-dpr-{dpr}'
export const FOLDERS = {
    ACTUAL: 'actual',
    DIFF: 'diff',
    TEMP_FULL_SCREEN: 'tempFullScreen',
    DEFAULT: {
        BASE: './__snapshots__/',
        SCREENSHOTS: '.tmp/',
    },
} as const
export const DEFAULT_SHADOW = {
    ADDRESS_BAR: 6,
    TOOL_BAR: 6,
}
export const DESKTOP = 'desktop'
export const NOT_KNOWN = 'not_known'
export const CUSTOM_CSS_ID = 'pic-css'
export const DEFAULT_RESIZE_DIMENSIONS: Required<ResizeDimensions> = {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
}
export const FULL_PAGE_SCROLL_TIMEOUT = 1500
export const DEFAULT_TABBABLE_OPTIONS = {
    circle: {
        backgroundColor: '#ff0000',
        borderColor: '#000',
        borderWidth: 1,
        fontColor: '#fff',
        fontFamily: 'Arial',
        fontSize: 10,
        size: 10,
        showNumber: true,
    },
    line: {
        color: '#000',
        width: 1,
    },
}
export const DEVICE_RECTANGLES: DeviceRectangles = {
    bottomBar: { y: 0, x: 0, width: 0, height: 0 },
    homeBar: { y: 0, x: 0, width: 0, height: 0 },
    leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
    rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
    statusBar: { y: 0, x: 0, width: 0, height: 0 },
    statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
    screenSize: { width: 0, height: 0 },
    viewport: { y: 0, x: 0, width: 0, height: 0 },
}

/**
 * Android and iOS offsets data
 */
const ANDROID_DEFAULT_OFFSETS = {
    STATUS_BAR: 24,
    ADDRESS_BAR: 56,
    TOOL_BAR: 48,
}
export const ANDROID_OFFSETS = {
    6: { ...ANDROID_DEFAULT_OFFSETS },
    7: { ...ANDROID_DEFAULT_OFFSETS },
    8: { ...ANDROID_DEFAULT_OFFSETS },
    9: { ...ANDROID_DEFAULT_OFFSETS },
    10: { ...ANDROID_DEFAULT_OFFSETS },
    11: { ...ANDROID_DEFAULT_OFFSETS },
    12: { ...ANDROID_DEFAULT_OFFSETS },
    13: { ...ANDROID_DEFAULT_OFFSETS },
    14: { ...ANDROID_DEFAULT_OFFSETS },
}

//=========================
// iPhone/iPad screen data
//=========================
// Height of screen in portrait mode, this is also what you get back
// when you:
//  - ask for the screen dimensions in WebdriverIO `getWindowSize`
//  - with `screen.height` in Safari devtools
//  - with http://whatismyscreenresolution.net/
// You can also check:
// - mydevice.io
// - https://useyourloaf.com/archives/
// - https://blisk.io/devices
// Corners come from:
// - https://github.com/kylebshr/ScreenCorners
//   One point is the equivalent of 1.333(3) pixels.
//   On the other hand, one pixel is the equivalent of 0.75 points.
//------------------------------------------------------------------
// iPhones
// 6/7/8/SE/2020:                                667
// 6/7/8/Plus:                                   736
// X|XS|11Pro|12Mini|13Mini (notch):             812
// 12|12Pro|13|13Pro|14 (notch):                 844
// 14Pro (notch):                                852
// XSMax|XR|11|11ProMax (notch):                 896
// 12ProMax|13ProMax|14Plus (notch):             926
// 14ProMax (notch):                             932
//------------------------------------------------------------------
// iPads
// 9.7|Air 1/2|Mini 1/2/3/4/2019|Pro 9.7:       1024
// 10.2 1/2019/2020/2021                        1080
// Pro 10.5/2017/2019:                          1112
// Mini 2021 (notch):                           1133
// Air 2020/2022 (notch):                       1180
// Pro 11 2018/2021 (notch):                    1194
// Pro 12.9 2017 (no notch)/2018/2021 (notch):  1366
//------------------------------------------------------------------
export const IOS_OFFSETS: IosOffsets = {
    IPHONE: {
        // 6/7/8/SE/2020
        667: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        // 6/7/8/Plus
        736: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        // X|XS|11Pro|12Mini|13Mini (notch)
        812: {
            LANDSCAPE: {
                SAFE_AREA: 44,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 300, y: 359, height: 9, width: 212 },
            },
            PORTRAIT: {
                SAFE_AREA: 44,
                STATUS_BAR: 44,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 118, y: 796, height: 9, width: 137 },
            },
        },
        // 12|12Pro|13|13Pro|14 (notch)
        844: {
            LANDSCAPE: {
                SAFE_AREA: 47,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 312, y: 375, height: 9, width: 220 },
            },
            PORTRAIT: {
                SAFE_AREA: 47,
                STATUS_BAR: 47,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 124, y: 829, height: 9, width: 143 },
            },
        },
        // 14Pro|15|16 (dynamic island)
        852: {
            LANDSCAPE: {
                SAFE_AREA: 59,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 316, y: 378, height: 9, width: 220 },
            },
            PORTRAIT: {
                SAFE_AREA: 59,
                STATUS_BAR: 59,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 125, y: 837, height: 9, width: 143 },
            },
        },
        // 16 Pro (dynamic island)
        874: {
            LANDSCAPE: {
                SAFE_AREA: 50,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 324, y: 389, height: 9, width: 226 },
            },
            PORTRAIT: {
                SAFE_AREA: 62,
                STATUS_BAR: 62,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 129, y: 861, height: 9, width: 152 },
            },
        },
        // XSMax|XR|11|11ProMax (notch)
        896: {
            LANDSCAPE: {
                SAFE_AREA: 48,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 331, y: 399, height: 9, width: 234 },
            },
            PORTRAIT: {
                SAFE_AREA: 48,
                STATUS_BAR: 48,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 131, y: 881, height: 9, width: 152 },
            },
        },
        // 12ProMax|13ProMax|14Plus (notch)
        926: {
            LANDSCAPE: {
                SAFE_AREA: 47,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 342, y: 413, height: 9, width: 242 },
            },
            PORTRAIT: {
                SAFE_AREA: 47,
                STATUS_BAR: 47,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 136, y: 911, height: 9, width: 156 },
            },
        },
        // 14ProMax|16 Plus (notch)
        932: {
            LANDSCAPE: {
                SAFE_AREA: 59,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 345, y: 415, height: 9, width: 242 },
            },
            PORTRAIT: {
                SAFE_AREA: 59,
                STATUS_BAR: 59,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 137, y: 917, height: 9, width: 158 },
            },
        },
        // 16 Pro Max (dynamic island)
        956: {
            LANDSCAPE: {
                SAFE_AREA: 50,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 355, y: 427, height: 9, width: 246 },
            },
            PORTRAIT: {
                SAFE_AREA: 62,
                STATUS_BAR: 62,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 142, y: 943, height: 9, width: 158 },
            },
        },
    },
    IPAD: {
        // 9.7|Air 1/2|Mini 1/2/3/4/2019|Pro 9.7
        1024: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        // 10.2 1/2019/2020/2021
        1080: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        // Pro 10.5/2017/2019
        1112: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        // With Home bar
        // Mini 2021
        1133: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 408, y: 729, height: 9, width: 318 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 234, y: 1118, height: 9, width: 276 },
            },
        },
        // Air 2020/2022
        1180: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 431, y: 805, height: 9, width: 318 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 272, y: 1165, height: 9, width: 276 },
            },
        },
        // Pro 11 2018/2021
        1194: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 438, y: 819, height: 9, width: 318 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 279, y: 1179, height: 9, width: 276 },
            },
        },
        // Pro 12.9 2017/2018/2021
        1366: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                // The iPad Pro (12.9 inch) (1st generation) has a status bar of 20px
                // STATUS_BAR: 24,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 508, y: 1009, height: 9, width: 350 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 352, y: 1351, height: 9, width: 320 },
            },
        },
    },
}

export const supportedIosBezelDevices = [
    // iPhones
    'iphonex',
    'iphonexs',
    'iphonexsmax',
    'iphonexr',
    'iphone11',
    'iphone11pro',
    'iphone11promax',
    'iphone12',
    'iphone12mini',
    'iphone12pro',
    'iphone12promax',
    'iphone13',
    'iphone13mini',
    'iphone13pro',
    'iphone13promax',
    'iphone14',
    'iphone14plus',
    'iphone14pro',
    'iphone14promax',
    'iphone15',
    // // iPads
    'ipadmini', // 1133
    'ipadair', // 1180
    'ipadpro11', // 1194
    'ipadpro129', // 1366
]

export const DEFAULT_TEST_CONTEXT: TestContext = {
    commandName: 'Could not be determined',
    instanceData: {
        app: 'Could not be determined',
        browser: {
            name: 'Could not be determined',
            version: 'Could not be determined',
        },
        deviceName: 'Could not be determined',
        isMobile: false,
        isAndroid: false,
        isIOS: false,
        platform: {
            name: 'Could not be determined',
            version: 'Could not be determined',
        },
    },
    framework: 'mocha',
    parent: 'Could not be determined',
    tag: 'Could not be determined',
    title: 'Could not be determined',
}
