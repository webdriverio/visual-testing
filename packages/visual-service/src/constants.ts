export const V6_CLIP_SELECTOR = '#root > :first-child:not(script):not(style)'
export const CLIP_SELECTOR = '#storybook-root > :first-child:not(script):not(style)'
export const NUM_SHARDS = 1
export const PAGE_OPTIONS_MAP: { [key: string]: string } = {
    'saveScreen': 'saveScreenOptions',
    'saveFullPageScreen': 'saveFullPageOptions',
    'saveTabbablePage': 'saveTabbableOptions',
    'checkScreen': 'checkScreenOptions',
    'checkFullPageScreen': 'checkFullPageOptions',
    'checkTabbablePage': 'checkTabbableOptions'
}
