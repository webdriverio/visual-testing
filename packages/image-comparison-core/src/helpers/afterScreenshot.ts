import { join } from 'node:path'
import logger from '@wdio/logger'
import hideScrollBars from '../clientSideScripts/hideScrollbars.js'
import removeElementFromDom from '../clientSideScripts/removeElementFromDom.js'
import { CUSTOM_CSS_ID } from './constants.js'
import { formatFileName, getAndCreatePath } from './utils.js'
import { saveBase64Image } from '../methods/images.js'
import type { AfterScreenshotOptions, ScreenshotOutput } from './afterScreenshot.interfaces.js'
import hideRemoveElements from '../clientSideScripts/hideRemoveElements.js'
import toggleTextTransparency from '../clientSideScripts/toggleTextTransparency.js'

const log = logger('@wdio/visual-service:@wdio/image-comparison-core:afterScreenshot')

/**
 * Methods that need to be executed after a screenshot has been taken
 * to set all back to the original state
 */
export default async function afterScreenshot(browserInstance: WebdriverIO.Browser, options: AfterScreenshotOptions): Promise<ScreenshotOutput> {
    const {
        actualFolder,
        base64Image,
        disableBlinkingCursor,
        disableCSSAnimation,
        enableLayoutTesting,
        fileName: fileNameOptions,
        filePath,
        hideElements,
        hideScrollBars: noScrollBars,
        isLandscape,
        isNativeContext,
        removeElements,
    } = options
    const path = getAndCreatePath(actualFolder, filePath)
    const fileName = formatFileName(fileNameOptions)

    await saveBase64Image(base64Image, join(path, fileName))

    const result = {
        devicePixelRatio: fileNameOptions.devicePixelRatio,
        fileName,
        isLandscape,
        path,
    }

    if (isNativeContext) {
        return result
    }

    if (noScrollBars) {
        await browserInstance.execute(hideScrollBars, !noScrollBars)
    }

    if ((hideElements && hideElements.length > 0) || (removeElements && removeElements.length > 0)) {
        try {
            await browserInstance.execute(hideRemoveElements, { hide: hideElements, remove: removeElements }, false)
        } catch (e) {
            log.warn(
                '\x1b[33m%s\x1b[0m',
                `
#####################################################################################
 WARNING:
 (One of) the elements that needed to be hidden or removed could not be found on the
 page and caused this error
 Error: ${e}
 We made sure the test didn't break.
#####################################################################################
`,
            )
        }
    }

    if (disableCSSAnimation || disableBlinkingCursor || browserInstance.isMobile) {
        await browserInstance.execute(removeElementFromDom, CUSTOM_CSS_ID)
    }

    if (enableLayoutTesting){
        await browserInstance.execute(toggleTextTransparency, !enableLayoutTesting)
    }

    return result
}
