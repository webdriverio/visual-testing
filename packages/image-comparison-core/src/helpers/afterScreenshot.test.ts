import { join } from 'node:path'
import logger from '@wdio/logger'
import { describe, it, expect, afterEach, beforeEach, vi } from 'vitest'
import afterScreenshot from './afterScreenshot.js'
import hideScrollBars from '../clientSideScripts/hideScrollbars.js'
import hideRemoveElements from '../clientSideScripts/hideRemoveElements.js'
import removeElementFromDom from '../clientSideScripts/removeElementFromDom.js'
import toggleTextTransparency from '../clientSideScripts/toggleTextTransparency.js'
import { CUSTOM_CSS_ID } from './constants.js'

const log = logger('test')

vi.mock('@wdio/logger', () => import(join(process.cwd(), '__mocks__', '@wdio/logger')))

vi.mock('../methods/images.js', () => ({
    saveBase64Image: vi.fn()
}))

vi.mock('./utils.js', () => ({
    getAndCreatePath: vi.fn(),
    formatFileName: vi.fn()
}))

import { saveBase64Image } from '../methods/images.js'
import { getAndCreatePath, formatFileName } from './utils.js'

describe('afterScreenshot', () => {
    const mockPath = '/mocked/path'
    const mockFileName = 'mocked-file-name.png'

    afterEach(() => {
        vi.clearAllMocks()
    })

    beforeEach(() => {
        vi.mocked(getAndCreatePath).mockReturnValue(mockPath)
        vi.mocked(formatFileName).mockReturnValue(mockFileName)
    })

    const createMockBrowserInstance = (
        mockExecuteFn = vi.fn().mockResolvedValue(''),
        customProperties: Partial<WebdriverIO.Browser> = {}
    ) => {
        return {
            execute: mockExecuteFn,
            ...customProperties
        } as unknown as WebdriverIO.Browser
    }
    const baseFilePath = {
        browserName: 'browserName',
        deviceName: 'deviceName',
        isMobile: false,
        savePerInstance: true,
    }
    const baseFileName = {
        browserName: 'browserName',
        browserVersion: 'browserVersion',
        deviceName: 'deviceName',
        devicePixelRatio: 2,
        formatImageName: '{tag}-{browserName}-{width}x{height}-dpr-{dpr}',
        isMobile: false,
        isTestInBrowser: true,
        logName: 'logName',
        name: 'name',
        outerHeight: 850,
        outerWidth: 1400,
        platformName: 'platformName',
        platformVersion: 'platformVersion',
        screenHeight: 900,
        screenWidth: 1440,
        tag: 'tag',
    }
    const createBaseOptions = (overrides = {}) => ({
        actualFolder: mockPath,
        base64Image: 'string',
        filePath: baseFilePath,
        fileName: baseFileName,
        hideScrollBars: false,
        isLandscape: false,
        isNativeContext: false,
        platformName: '',
        ...overrides,
    })

    it('should be able to return the ScreenshotOutput with default options', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const options = createBaseOptions({
            disableBlinkingCursor: false,
            disableCSSAnimation: false,
            hideScrollBars: true,
            hideElements: [<HTMLElement>(<any>'<div></div>')],
            removeElements: [<HTMLElement>(<any>'<div></div>')],
        })

        const result = await afterScreenshot(mockBrowserInstance, options)

        expect(vi.mocked(getAndCreatePath)).toHaveBeenCalledWith(mockPath, options.filePath)
        expect(vi.mocked(formatFileName)).toHaveBeenCalledWith(options.fileName)
        expect(vi.mocked(saveBase64Image)).toHaveBeenCalledWith(options.base64Image, join(mockPath, mockFileName))

        expect(result).toMatchSnapshot()
    })

    it('should handle native context and skip browser operations', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const options = createBaseOptions({
            disableBlinkingCursor: true,
            disableCSSAnimation: true,
            enableLayoutTesting: true,
            fileName: {
                ...baseFileName,
                devicePixelRatio: 1.5,
                outerHeight: 600,
                outerWidth: 800,
                screenHeight: 700,
                screenWidth: 900,
            },
            hideScrollBars: true,
            isLandscape: true,
            isNativeContext: true,
            hideElements: [<HTMLElement>(<any>'<div></div>')],
            platformName: 'iOS',
            removeElements: [<HTMLElement>(<any>'<div></div>')],
        })

        const result = await afterScreenshot(mockBrowserInstance, options)

        expect(vi.mocked(getAndCreatePath)).toHaveBeenCalledWith(mockPath, options.filePath)
        expect(vi.mocked(formatFileName)).toHaveBeenCalledWith(options.fileName)
        expect(vi.mocked(saveBase64Image)).toHaveBeenCalledWith(options.base64Image, join(mockPath, mockFileName))
        expect(mockBrowserInstance.execute).not.toHaveBeenCalled()
        expect(result).toMatchSnapshot()
    })

    it('should handle layout testing with enableLayoutTesting', async () => {
        const mockExecute = vi.fn().mockResolvedValue('')
        const mockBrowserInstance = createMockBrowserInstance(mockExecute)
        const options = createBaseOptions({
            enableLayoutTesting: true,
        })
        const result = await afterScreenshot(mockBrowserInstance, options)

        expect(vi.mocked(getAndCreatePath)).toHaveBeenCalledWith(mockPath, options.filePath)
        expect(vi.mocked(formatFileName)).toHaveBeenCalledWith(options.fileName)
        expect(vi.mocked(saveBase64Image)).toHaveBeenCalledWith(options.base64Image, join(mockPath, mockFileName))
        expect(mockExecute).toHaveBeenCalledWith(toggleTextTransparency, false)
        expect(result).toMatchSnapshot()
    })

    it('should handle mobile platform and remove custom CSS', async () => {
        const mockExecute = vi.fn().mockResolvedValue('')
        const mockBrowserInstance = createMockBrowserInstance(mockExecute, { isMobile: true })
        const options = createBaseOptions({
            disableBlinkingCursor: false,
            disableCSSAnimation: false,
            filePath: {
                ...baseFilePath,
                isMobile: true,
            },
            fileName: {
                ...baseFileName,
                isMobile: true,
                platformName: 'Android',
            },
            platformName: 'Android',
        })
        const result = await afterScreenshot(mockBrowserInstance, options)

        expect(vi.mocked(getAndCreatePath)).toHaveBeenCalledWith(mockPath, options.filePath)
        expect(vi.mocked(formatFileName)).toHaveBeenCalledWith(options.fileName)
        expect(vi.mocked(saveBase64Image)).toHaveBeenCalledWith(options.base64Image, join(mockPath, mockFileName))
        expect(mockExecute).toHaveBeenCalledWith(removeElementFromDom, CUSTOM_CSS_ID)
        expect(result).toMatchSnapshot()
    })

    it('should handle hide/remove elements with error handling', async () => {
        const mockExecute = vi.fn().mockRejectedValueOnce(new Error('Element not found'))
        const mockBrowserInstance = createMockBrowserInstance(mockExecute)
        const hideElements = [<HTMLElement>(<any>'<div></div>')]
        const removeElements = [<HTMLElement>(<any>'<div></div>')]
        const options = createBaseOptions({
            hideElements,
            removeElements,
        })
        const result = await afterScreenshot(mockBrowserInstance, options)

        expect(vi.mocked(getAndCreatePath)).toHaveBeenCalledWith(mockPath, options.filePath)
        expect(vi.mocked(formatFileName)).toHaveBeenCalledWith(options.fileName)
        expect(vi.mocked(saveBase64Image)).toHaveBeenCalledWith(options.base64Image, join(mockPath, mockFileName))
        expect(mockExecute).toHaveBeenCalledWith(hideRemoveElements, { hide: hideElements, remove: removeElements }, false)
        expect(log.warn).toHaveBeenCalledTimes(1)
        expect(vi.mocked(log.warn).mock.calls[0]).toMatchSnapshot()
        expect(result).toMatchSnapshot()
    })

    it('should handle hideScrollBars when hideScrollBars is true', async () => {
        const mockExecute = vi.fn().mockResolvedValue('')
        const mockBrowserInstance = createMockBrowserInstance(mockExecute)
        const options = createBaseOptions({
            hideScrollBars: true,
        })
        const result = await afterScreenshot(mockBrowserInstance, options)
        expect(vi.mocked(getAndCreatePath)).toHaveBeenCalledWith(mockPath, options.filePath)
        expect(vi.mocked(formatFileName)).toHaveBeenCalledWith(options.fileName)
        expect(vi.mocked(saveBase64Image)).toHaveBeenCalledWith(options.base64Image, join(mockPath, mockFileName))
        expect(mockExecute).toHaveBeenCalledWith(hideScrollBars, false)
        expect(result).toMatchSnapshot()
    })

    it('should skip hide/remove elements when both are empty arrays', async () => {
        const mockExecute = vi.fn().mockResolvedValue('')
        const mockBrowserInstance = createMockBrowserInstance(mockExecute)
        const options = createBaseOptions({
            hideElements: [],
            removeElements: [],
        })
        const result = await afterScreenshot(mockBrowserInstance, options)

        expect(vi.mocked(getAndCreatePath)).toHaveBeenCalledWith(mockPath, options.filePath)
        expect(vi.mocked(formatFileName)).toHaveBeenCalledWith(options.fileName)
        expect(vi.mocked(saveBase64Image)).toHaveBeenCalledWith(options.base64Image, join(mockPath, mockFileName))
        expect(mockExecute).not.toHaveBeenCalledWith(hideRemoveElements, expect.any(Object), false)
        expect(result).toMatchSnapshot()
    })

    it('should skip hide/remove elements when both are falsy', async () => {
        const mockExecute = vi.fn().mockResolvedValue('')
        const mockBrowserInstance = createMockBrowserInstance(mockExecute)
        const options = createBaseOptions({
            hideElements: undefined,
            removeElements: null,
        })
        const result = await afterScreenshot(mockBrowserInstance, options)

        expect(vi.mocked(getAndCreatePath)).toHaveBeenCalledWith(mockPath, options.filePath)
        expect(vi.mocked(formatFileName)).toHaveBeenCalledWith(options.fileName)
        expect(vi.mocked(saveBase64Image)).toHaveBeenCalledWith(options.base64Image, join(mockPath, mockFileName))
        expect(mockExecute).not.toHaveBeenCalledWith(hideRemoveElements, expect.any(Object), false)
        expect(result).toMatchSnapshot()
    })

    it('should handle only hideElements with length > 0', async () => {
        const mockExecute = vi.fn().mockResolvedValue('')
        const mockBrowserInstance = createMockBrowserInstance(mockExecute)
        const hideElements = [<HTMLElement>(<any>'<div></div>')]
        const options = createBaseOptions({
            hideElements,
            removeElements: [],
        })
        const result = await afterScreenshot(mockBrowserInstance, options)

        expect(vi.mocked(getAndCreatePath)).toHaveBeenCalledWith(mockPath, options.filePath)
        expect(vi.mocked(formatFileName)).toHaveBeenCalledWith(options.fileName)
        expect(vi.mocked(saveBase64Image)).toHaveBeenCalledWith(options.base64Image, join(mockPath, mockFileName))
        expect(mockExecute).toHaveBeenCalledWith(hideRemoveElements, { hide: hideElements, remove: [] }, false)
        expect(result).toMatchSnapshot()
    })

    it('should handle only removeElements with length > 0', async () => {
        const mockExecute = vi.fn().mockResolvedValue('')
        const mockBrowserInstance = createMockBrowserInstance(mockExecute)
        const removeElements = [<HTMLElement>(<any>'<div></div>')]
        const options = createBaseOptions({
            hideElements: null,
            removeElements,
        })
        const result = await afterScreenshot(mockBrowserInstance, options)

        expect(vi.mocked(getAndCreatePath)).toHaveBeenCalledWith(mockPath, options.filePath)
        expect(vi.mocked(formatFileName)).toHaveBeenCalledWith(options.fileName)
        expect(vi.mocked(saveBase64Image)).toHaveBeenCalledWith(options.base64Image, join(mockPath, mockFileName))
        expect(mockExecute).toHaveBeenCalledWith(hideRemoveElements, { hide: null, remove: removeElements }, false)
        expect(result).toMatchSnapshot()
    })
})
