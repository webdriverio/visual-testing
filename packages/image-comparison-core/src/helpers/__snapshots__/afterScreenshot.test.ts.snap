// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`afterScreenshot > should be able to return the ScreenshotOutput with default options 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "mocked-file-name.png",
  "isLandscape": false,
  "path": "/mocked/path",
}
`;

exports[`afterScreenshot > should handle hide/remove elements with error handling 1`] = `
[
  "[33m%s[0m",
  "
#####################################################################################
 WARNING:
 (One of) the elements that needed to be hidden or removed could not be found on the
 page and caused this error
 Error: Error: Element not found
 We made sure the test didn't break.
#####################################################################################
",
]
`;

exports[`afterScreenshot > should handle hide/remove elements with error handling 2`] = `
{
  "devicePixelRatio": 2,
  "fileName": "mocked-file-name.png",
  "isLandscape": false,
  "path": "/mocked/path",
}
`;

exports[`afterScreenshot > should handle hideScrollBars when hideScrollBars is true 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "mocked-file-name.png",
  "isLandscape": false,
  "path": "/mocked/path",
}
`;

exports[`afterScreenshot > should handle layout testing with enableLayoutTesting 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "mocked-file-name.png",
  "isLandscape": false,
  "path": "/mocked/path",
}
`;

exports[`afterScreenshot > should handle mobile platform and remove custom CSS 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "mocked-file-name.png",
  "isLandscape": false,
  "path": "/mocked/path",
}
`;

exports[`afterScreenshot > should handle native context and skip browser operations 1`] = `
{
  "devicePixelRatio": 1.5,
  "fileName": "mocked-file-name.png",
  "isLandscape": true,
  "path": "/mocked/path",
}
`;

exports[`afterScreenshot > should handle only hideElements with length > 0 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "mocked-file-name.png",
  "isLandscape": false,
  "path": "/mocked/path",
}
`;

exports[`afterScreenshot > should handle only removeElements with length > 0 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "mocked-file-name.png",
  "isLandscape": false,
  "path": "/mocked/path",
}
`;

exports[`afterScreenshot > should skip hide/remove elements when both are empty arrays 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "mocked-file-name.png",
  "isLandscape": false,
  "path": "/mocked/path",
}
`;

exports[`afterScreenshot > should skip hide/remove elements when both are falsy 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "mocked-file-name.png",
  "isLandscape": false,
  "path": "/mocked/path",
}
`;
